crates/scheduler/src/bin/hypha-scheduler.rs

```diff
@@ -331,42 +336,78 @@ async fn run(config: ConfigWithMetadata<Config>) -> Result<()> {
 
     let tracker_task = tokio::spawn(slice_tracker.run().await.expect("network ready"));
 
-    let status_handle = if allocated_workers.len() > 1 && allocated_parameter_servers.len() == 1 {
+    if allocated_workers.len() > 1 && allocated_parameter_servers.len() == 1 {
+        let job_id = Uuid::new_v4();
+        let batch_sizes = [22, 12];
+        let samples_per_update = 220;
+        let diloco_rounds = 4;
+
         let worker_ids = allocated_workers
             .iter()
             .map(|w| w.peer_id())
             .collect::<Vec<_>>();
         let parameter_server = &mut allocated_parameter_servers[0];
+        // TODO compute true batch sizes
+        let run_tracker = Arc::new(Mutex::new(ProgressTracker::<RunningMean>::new(
+            parameter_server.peer_id(),
+            samples_per_update,
+            diloco_rounds,
+        )));
+        let batch_scheduler = BatchScheduler::new(
+            network.clone(),
+            BasicSimulation::project,
+            run_tracker.clone(),
+            job_id,
+        );
+        let (stream, batch_scheduler_handle) = batch_scheduler
+            .run()
+            .await
+            .expect("Batch Scheduler started");
 
         let model_files = vec!["config.json".to_string(), "model.safetensors".to_string()];
 
-        let w1_rx = allocated_workers[0]
+        let _w1_rx = allocated_workers[0]
             .dispatch(JobSpec {
+                job_id,
                 executor: get_executor_with_dataset(
                     model_files.clone(),
                     peer_id,
                     dataset.to_string(),
                     parameter_server.peer_id(),
+                    batch_sizes[0],
                 ),
             })
             .await
             .into_diagnostic()?;
+        run_tracker
+            .lock()
+            .await
+            .worker_tracker
+            .add_worker(allocated_workers[0].peer_id(), batch_sizes[0]);
```

Comment:
We should add workers before dispatching a job to avoid races. Even better to add them before running the batch scheduler. If messages arrive in the batch scheduler, it'll try to find the worker in `run_tracker`. In worst case scenarios, first messages arrive before we've added the worker to `run_tracker`. Right now this would crash the scheduler.